text/profiles/0239-profiles-data-model.md

```diff
@@ -0,0 +1,1601 @@
+  // The set of samples recorded in this profile.
+  repeated Sample sample = 2;
+  // Mapping from address ranges to the image/binary/library mapped
+  // into that address range.  mapping[0] will be the main binary.
+  repeated Mapping mapping = 3;
+  // Locations referenced by samples via location_indices.
+  repeated Location location = 4;
+  // Array of locations referenced by samples.
+  repeated int64 location_indices = 15;
+  // Functions referenced by locations.
+  repeated Function function = 5;
+  // A common table for strings referenced by various messages.
+  // string_table[0] must always be "".
+  // Lookup table for attributes.
+  repeated opentelemetry.proto.common.v1.KeyValue attribute_table = 16;
+  // Represents a mapping between Attribute Keys and Units.
+  repeated AttributeUnit attribute_units = 17;
+  // Lookup table for links.
+  repeated Link link_table = 18;
+  repeated string string_table = 6;
+  // frames with Function.function_name fully matching the following
+  // regexp will be dropped from the samples, along with their successors.
+  int64 drop_frames = 7;   // Index into string table.
+  // frames with Function.function_name fully matching the following
+  // regexp will be kept, even if it matches drop_frames.
+  int64 keep_frames = 8;  // Index into string table.
+
+  // The following fields are informational, do not affect
+  // interpretation of results.
+
+  // Time of collection (UTC) represented as nanoseconds past the epoch.
+  int64 time_nanos = 9;
+  // Duration of the profile, if a duration makes sense.
+  int64 duration_nanos = 10;
+  // The kind of events between sampled occurrences.
+  // e.g [ "cpu","cycles" ] or [ "heap","bytes" ]
+  ValueType period_type = 11;
+  // The number of events between sampled occurrences.
+  int64 period = 12;
+  // Free-form text associated with the profile. The text is displayed as is
+  // to the user by the tools that read profiles (e.g. by pprof). This field
+  // should not be used to store any machine-readable information, it is only
+  // for human-friendly content. The profile must stay functional if this field
+  // is cleaned.
+  repeated int64 comment = 13; // Indices into string table.
+  // Index into the string table of the type of the preferred sample
+  // value. If unset, clients should default to the last sample value.
+  int64 default_sample_type = 14;
+}
+
+// A pointer from a profile Sample to a trace Span.
+// Connects a profile sample to a trace span, identified by unique trace and span IDs.
+message Link {
+  // A unique identifier of a trace that this linked span is part of. The ID is a
+  // 16-byte array.
+  bytes trace_id = 1;
+
+  // A unique identifier for the linked span. The ID is an 8-byte array.
+  bytes span_id = 2;
+}
+
+// Specifies the method of aggregating metric values, either DELTA (change since last report)
+// or CUMULATIVE (total since a fixed start time).
+enum AggregationTemporality {
```

Comment:
FWIW I never reviewed this enum and all the extensive comments in it in details. I would probably prefer if we leave it out of this initial PR and add later, reviewing the concept and the comments holistically.

Reply:
I know the comments are quite long, but the concept is realtively simple — there's two types of aggregation temporality, `delta` and `cumulative`. 
* `delta` is the default type that you'd see for example in CPU profiles where you reset all values for each new profile you send
* `cumulative` is a profile where the values always go up, e.g Go `alloc_objects` profiles.

This concept is used in [metrics spec as well](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L249-L318), and I just copied the type along with the comments.

Regarding removing it vs keeping it I would vote for keeping it here, because this way we can start experimenting with it. And if it turns out we decide to keep it, that's one less thing we'll need to change. And if we decide to remove it or change it we'll still have that option. But I'm happy to discuss this further live at the Profiling SIG meeting.

Reply:
That's a good point. For reference, The way it's handled in metrics is by using two separate types of messages: [gauges and sums](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L193-L217), where gauges don't have aggregation temporality, but sums do, so you end up with 3 choices:
* gauge
* sum delta
* sum cumulative

I imagine this is not going to work in our case — let me think about this some more.